### DISCUSSION TITLE
Can't install on Windows

### DISCUSSION THREAD

**Discussion by Mouthless-Stoat** [lumen-oss/rocks.nvim#407] [Q&A] [answered]
I have try all 3 method of installing:
- Manually
- Script
- Bootstrap

All seem to fail with different error. First the bootstrap always fail at the source command so I can't really see what is wrong. Using the script I always seem to get stuck around configuring`luarocks` it always `Missing Lua 5.1 Interpreter` or on installing. And manually the furthest I got is to actually get `rocks.nvim` to install but `toml-edit` seem to be missing.

I try several way of getting both `lua` and `luarocks`:
- `scoop install luarocks lua`
- `scoop install lua-for-windows`
- The `LuaForWindow.msi` from the [github repo](https://github.com/rjpcomputing/luaforwindows/releases/tag/v5.1.5-52)
- The binaries for both `luarocks` and `lua` from their website ([`luarocks`](https://github.com/luarocks/luarocks/wiki/Download) and [`lua`](https://sourceforge.net/projects/luabinaries/files/5.1.5/))
- Using `hererocks`

Apart from `scoop install lua` which download `Lua v5.4.2` all other download are `Lua v5.1.5`. The most promising approach seem to be with `hererocks` and using the manual method of install. 

When I check the log for this approach it fails on making some `symlink`. It seem to try and make a `symlink` between `rocks/share/lua/5.1` with `rocks/rock_rtp/lua` and get `operation not permitted`, but I don't think it matter because `toml-edit` seem to not be install(?), but i do see it when i try to manually install `rocks.nvim` again

```
[START][2024-06-23 01:46:41] rocks.nvim logging initiated
ERROR | 2024-06-23 01:46:41 | ...ta\Local\nvim-data/rocks/share/lua/5.1/rocks\adapter.lua:40 | Error creating symlink directory: lua (EPERM: operation not permitted: C:\Users\*****\AppData\Local\nvim-data/rocks/share/lua/5.1 -> C:\Users\****\AppData\Local\nvim-data/rocks/rocks_rtp/lua)
ERROR | 2024-06-23 01:46:42 | ...a\Local\nvim-data\rocks\share\lua\5.1\rocks\luarocks.lua:64 | luarocks CLI FAILED
ERROR | 2024-06-23 01:46:42 | ...a\Local\nvim-data\rocks\share\lua\5.1\rocks\luarocks.lua:65 | Failed to invoke luarocks at luarocks: ...ps\neovim\current\share\nvim\runtime/lua/vim/_system.lua:244: ENOENT: no such file or directory
ERROR | 2024-06-23 01:46:42 | ...a\Local\nvim-data\rocks\share\lua\5.1\rocks\luarocks.lua:64 | luarocks CLI FAILED
ERROR | 2024-06-23 01:46:42 | ...a\Local\nvim-data\rocks\share\lua\5.1\rocks\luarocks.lua:65 | Failed to invoke luarocks at luarocks: ...ps\neovim\current\share\nvim\runtime/lua/vim/_system.lua:244: ENOENT: no such file or directory
ERROR | 2024-06-23 01:46:42 | ...a\Local\nvim-data\rocks\share\lua\5.1\rocks\luarocks.lua:64 | luarocks CLI FAILED
ERROR | 2024-06-23 01:46:42 | ...a\Local\nvim-data\rocks\share\lua\5.1\rocks\luarocks.lua:65 | Failed to invoke luarocks at luarocks: ...ps\neovim\current\share\nvim\runtime/lua/vim/_system.lua:244: ENOENT: no such file or directory

```

I also found #166 and try a few thing from there (mainly using `hererocks` instead of `scoop`) but still nothing

**Comment by mrcjkb** [comment]
Hey 👋

Windows support isn't fully implemented yet, so the bootstrap and installer scripts won't work unless you have the unix tools it depends on installed).

The process on Windows is still very finicky and often requires a lot of manual troubleshooting.

- `ENOENT: no such file or directory` means that the `luarocks` command is not available (likely because it's not on the `PATH`).
- `EPERM: operation not permitted` means the neovim process doesn't have permissions to create the symlinks.

Whatever tool you use to install luarocks, make sure you install the latest version (3.11.1 as of writing this reply).

**Reply by Mouthless-Stoat** [reply]
I try running `luarocks --version` which yield `luarocks 3.11.1` and `lua -v` yield `Lua 5.1.5`. So `luarocks` and `lua` seem to be installed correctly. As for the symlink issue I launch `nvim` with admin and nothing seem to change. Although I did see that it create a file symlink instead of a dir symlink. So I modified the `adapter.lua` a bit to tell it to create dir symlink and still nothing change.

Still the issue is it can't find `toml-edit`
![image](https://github.com/nvim-neorocks/rocks.nvim/assets/89868169/67347af9-6074-43f1-8f75-ca6c74f4f7f2)

**Reply by Mouthless-Stoat** [reply]
Seem to get it to install after manually putting the `toml-edit.dll` file where it is expecting it. When I run `:Rocks install telescope` it seem to install just fine but nothing it added to the `rock,toml` but I assume this is how `rock.nvim` work? But then it error about missing `Lua 5.1` whenever I try to install stuff via `rock.toml` and `:Rocks sync` (It also uninstall itself?)

`rock.toml`:
```
[plugins]
"telescope.nvim" = {version = "0.8.4"}
```

**Reply by mrcjkb** [reply]
> I try running luarocks --version which yield luarocks 3.11.1 and lua -v yield Lua 5.1.5. So luarocks and lua seem to be installed correctly.

That doesn't necessarily mean your environment is set up so that Neovim can find them (see [this comment](https://github.com/nvim-neorocks/rocks.nvim/issues/166#issuecomment-1966946445)).

> Still the issue is it can't find toml-edit

Based on your error message it looks like you haven't set up the `package.cpath` correctly. On Windows, you need to make sure it also searches for `.dll` files.

> When I run :Rocks install telescope it seem to install just fine but nothing it added to the rock,toml but I assume this is how rock.nvim work?

No, that's not now it's supposed to work. It's likely related to your issues with toml-edit.

> (It also uninstall itself?)

This is because you manually created a rocks.toml without a rocks.nvim entry.

**Reply by Mouthless-Stoat** [reply]
> That doesn't necessarily mean your environment is set up so that Neovim can find them (see https://github.com/nvim-neorocks/rocks.nvim/issues/166#issuecomment-1966946445).

Running `:=vim.system({"luarocks", "--version"}):wait()` and `:=vim.system({"lua", "-v"}):wait()` seem to yield similar result so `nvim` is definitely detecting them correctly

> Based on your error message it looks like you haven't set up the package.cpath correctly. On Windows, you need to make sure it also searches for .dll files.

My `LUA_CPATH` was just not set at all so after setting it correctly `nvim` seem to launch with no error. But `toml-edit` seem to just clear the `rocks.toml` file as after `:Rocks install telescope.nvim` the file just clear even when there was something in it. When running `:Rocks sync` it seem to read correctly just not write

**Reply by mrcjkb** [reply, marked as answer]
See [this wiki article](https://github.com/nvim-neorocks/rocks.nvim/wiki/Installing-rocks.nvim-manually,-without-the-installation-script) for how to set up the paths when installing manually.

> When running :Rocks sync it seem to read correctly just not write

This looks like your Neovim installation has permissions issues (just like your symlink errors).

After correctly configuring the paths, try deleting the rocks.toml and let rocks.nvim create it for you. 

If you still have issues with permissions, I'm afraid I don't know how to help you, since I don't use Windows.

**Reply by Mouthless-Stoat** [reply]
I been launching `nvim` with admin permission and follow the wiki page to install it. Ig I gotta wait till Windows stabilize a bit more. Thank you so much for helping tho.
